### PR TITLE
Fix: Read replicas connection settings replaced by master settings

### LINKED ISSUE #1963: Read replicas connection settings replaced by "master" settings
**Describe the bug**
Read replicas connection settings replaced by "master" settings, when connection settings specified as object.

**To Reproduce**
You may use example provided in the manual: https://mikro-orm.io/docs/read-connections
In this case replicas host and user parameter values will be replaced by top-level (master) configuration parameter values.

This issue occurs because `Connection` class instance `options` property value is replaced by the corresponding value from top-level config in `getConnectionOptions()`.

It works as expected when replica and master connection settings are specified in the `clientUrl` parameter.

```
  getConnectionOptions(): ConnectionConfig {
    const ret: ConnectionConfig = {};
    const url = new URL(this.options!.clientUrl || this.config.getClientUrl());
    this.options!.host = ret.host = this.config.get('host', url.hostname);
    this.options!.port = ret.port = this.config.get('port', +url.port);
    this.options!.user = ret.user = this.config.get('user', url.username);
    this.options!.password = ret.password = this.config.get('password', url.password);
    this.options!.dbName = ret.database = this.config.get('dbName', url.pathname.replace(/^\//, ''));

    return ret;
  }
```
See:
https://github.com/mikro-orm/mikro-orm/blob/091e60d5636959477176d851ef112ab53f1c6609/packages/core/src/connections/Connection.ts#L64

**Versions**

| Dependency | Version |
| - | - |
| node | 14.17.0 |
| typescript | 4.1.2 |
| mikro-orm | 4.5.5 |
| your-driver | postgresql |

### PR DESCRIPTION
**Pull request by rbrishabh**
fixes #1963 

Hey @B4nan! How does this look?
let me know if this needs any changes!
Thanks

### CODE REVIEW

File: packages/core/src/connections/Connection.ts [hunk 1]
```diff
@@ -64,11 +64,11 @@ export abstract class Connection {
...
-    this.options!.host = ret.host = this.config.get('host', url.hostname);
-    this.options!.port = ret.port = this.config.get('port', +url.port);
-    this.options!.user = ret.user = this.config.get('user', url.username);
-    this.options!.password = ret.password = this.config.get('password', url.password);
-    this.options!.dbName = ret.database = this.config.get('dbName', url.pathname.replace(/^\//, ''));
+    this.options!.host = ret.host = (this.options!.host || this.config.get('host', url.hostname));
+    this.options!.port = ret.port = (this.options!.port || this.config.get('port', +url.port));
+    this.options!.user = ret.user = (this.options!.user || this.config.get('user', url.username));
+    this.options!.password = ret.password = (this.options!.password || this.config.get('password', url.password));
+    this.options!.dbName = ret.database = (this.options!.dbName || this.config.get('dbName', url.pathname.replace(/^\//, '')));
```
Comment (B4nan) [inline packages/core/src/connections/Connection.ts:71]: let's use `??` instead

```suggestion
    this.options!.host = ret.host = (this.options!.host ?? this.config.get('host', url.hostname));
    this.options!.port = ret.port = (this.options!.port ?? this.config.get('port', +url.port));
    this.options!.user = ret.user = (this.options!.user ?? this.config.get('user', url.username));
    this.options!.password = ret.password = (this.options!.password ?? this.config.get('password', url.password));
    this.options!.dbName = ret.database = (this.options!.dbName ?? this.config.get('dbName', url.pathname.replace(/^\//, '')));
```

**Review by B4nan — CHANGES_REQUESTED** [review]
i guess that should make it work, would be great to add a test for this to ensure it really helps

**Comment by rbrishabh** [comment]
ok! I can do that! the test goes in here, right? `tests/Connection.test.ts`?

**Comment by B4nan** [comment]
replicas are created in the constructor of `AbstractSqlDriver`, so we should test it from there. you could just initialize the ORM with given config that uses replicas (and has different connection options for them than the master db), mock the `getConnectionOptions` method of given connection class via `jest.spyOn` and check the params that were used.

**Comment by rbrishabh** [comment]
got it! will get to this soon!

**Comment by rbrishabh** [comment]
hi! I have a busy schedule and cant keep up with this now! Please feel free to continue it from here. I can close it for now

**Comment by ShinteiMai** [comment]
@rbrishabh Hey, I would be glad to take over this issue. I'll continue from the work that you've done

**Comment by rbrishabh** [comment]
Perfect! Go for it!

**Comment by rbrishabh** [comment]
I've reopened the PR for you :)

**Comment by ShinteiMai** [comment]
cc @rbrishabh I don't have write permission to your fork, so i forked my own repo then remote pulled your branch, here's the new PR https://github.com/mikro-orm/mikro-orm/pull/2081

**Pull request closed without merging**